pkg/dns/dns.go

```diff
@@ -147,13 +164,21 @@ func (kd *KubeDNS) updateConfig(nextConfig *config.Config) {
 		var nameServers []string
 		for _, nameServer := range nextConfig.UpstreamNameservers {
 			if ip, port, err := util.ValidateNameserverIpAndPort(nameServer); err != nil {
-				glog.V(1).Infof("Invalid nameserver %q: %v", nameServer, err)
+				glog.Errorf("Invalid nameserver %q: %v", nameServer, err)
+				if len(kd.SkyDNSConfig.Nameservers) == 0 {
+					// Fall back to resolv.conf on initialization failed.
+					kd.SkyDNSConfig.Nameservers = kd.loadDefaultNameserver()
```

Comment:
I think we should fall through here --- append whatever valid nameservers are passed in, and reuse the 0 length detection logic below:
```golang
if ip, port, err := util.ValidateNameserverIpAndPort(nameServer); err != nil {
   glog.Errorf("Invalid nameserver %q: %v", nameServer, err)
} else {
   nameServers = append(nameServers, net.JoinHostPort(ip, port))
}
...
```

Reply:
The idea is, we don't want to pollute kd.config with invalid input. (But the truth is, kd.config.UpstreamNameservers are never used so it's okay to be polluted)

Reply:
Sounds good, though having duplicated logic here seems bad. What about:
```golang
	if kd.SkyDNSConfig != nil {
		var nameServers []string
		for _, nameServer := range nextConfig.UpstreamNameservers {
			if ip, port, err := util.ValidateNameserverIpAndPort(nameServer); err != nil {
				glog.Errorf("Invalid nameserver %q: %v", nameServer, err)
				// Don't populate nameServers if invalid input is detected.
				nameServers = []string{}
				break
			} else {
				nameServers = append(nameServers, net.JoinHostPort(ip, port))
			}
		}
		if len(nameServers) == 0 {
			...
```

Reply:
Sorry what I wrote is wrong. I got what you were saying.

Reply:
But we can still do the partially update and leave kd.config untouched. Do you want me to do that?
I am kind of against this since kd.SkyDNSConfig is partially updated and kd.config is not which might be confusing in some extreme cases.